docs/tutorial/durham_reframe_tutorial.md

```diff
@@ -0,0 +1,409 @@
+- `valid_prog_environs` for what compilers this test can build with. More on it later.
+- `sourcepath` for source file in a single source test. More on build systems later.
+- Could add `sourcesdir` but it defaults to `src/`
+
+```python
+    valid_systems = ['*']
+    valid_prog_environs = ['*']
+    sourcepath = 'hello.c'
+```
+
+----
+
+### Add sanity check
+
+- ReFrame, by default, makes no assumption about whether a test is successful or not.
+- A test must provide a validation function
+- ReFrame provides a rich set of utility functions that help matching patterns and extract values from the test’s output
+- Here we match a string from stdout
+
+```python
+    @sanity_function
+    def assert_hello(self):
+        return sn.assert_found(r'Hello, World\!', self.stdout)
+```
+
+----
+
+## Builting programming environment
+
+- `reframe --show-config`
+- Builtin programming environment uses `cc` to compile
+
+---
+
+## Configuring ReFrame for HPC systems (Cosma)
+> In ReFrame, all the details of the various interactions of a test with the system environment are handled transparently and are set up in its configuration file.
+- [Configuration](https://reframe-hpc.readthedocs.io/en/v4.5.2/tutorial_basics.html#more-of-hello-world)
+  - Set accounting parameters with
+    - `'access': ['--partition=bluefield1', '--account=do009'],`
+  - Create at least one programming environment to set compilers
+    - `-p` flag filters tests by programming environment
+  - Scheduler to run on compute nodes
+    - Add `time_limit = 1m` to ReFrame tests to run on Cosma
+    - Set from command line with `-S time_limit='1m'`
+
+----
+
+```python
+site_configuration = {
+    'systems' : [
+        {
+            'name': 'cosma',
+            'descr': 'Cosma for performance workshop',
+            'hostnames': ['login[0-9][a-z].pri.cosma[0-9].alces.network'],
+            'modules_system': 'tmod4',
+            'partitions': [
+                {
+                    'name': 'compute-node',
+                    'scheduler': 'slurm',
+                    'launcher': 'mpiexec',
+                    'environs': ['gnu'],
+                    'access': ['--partition=bluefield1', '--account=do009'],
+                }
+            ]
+        }
+    ],
+    'environments': [
+        {
+            'name': 'gnu',
+            'modules': ['gnu_comp', 'openmpi'],
+            'cc': 'mpicc',
+            'cxx': 'mpic++',
+            'ftn': 'mpif90'
+        },
+    ]
+}
+
+```
+
+---
+
+## Performance tests
+
+Performance tests capture data in performance variables. For simplicity, we use the [STREAM benchmark](https://github.com/jeffhammond/STREAM) as an example. It is the de facto memory bandwidth benchmark.
+
+To record the performance of the benchmark, ReFrame should extract a figure of merit from the output of the test. A function decorated with the `@performance_function` decorator extracts or computes a performance metric from the test’s output.
```

Comment:
Remove this paragraph, it's duplicated in the "Add Performance Pattern Check" section below